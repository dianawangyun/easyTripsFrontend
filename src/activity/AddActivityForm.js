import React, { useState, useContext } from "react";
import UserContext from "../auth/UserContext";

import usePlacesAutocomplete, {
    getGeocode,
    getLatLng,
} from "use-places-autocomplete";
import {
    Combobox,
    ComboboxInput,
    ComboboxPopover,
    ComboboxList,
    ComboboxOption,
} from "@reach/combobox";
import { TextField } from "@material-ui/core";
import "@reach/combobox/styles.css";

function AddActivityForm({
    tripId,
    setIsAdding,
    addActivity,
    setSelectLocation,
}) {
    const { currLocation } = useContext(UserContext);
    const [formData, setFormData] = useState({
        tripId: tripId,
        activityName: "",
    });
    const [formErrors, setFormErrors] = useState([]);

    const {
        ready,
        value,
        suggestions: { status, data },
        setValue,
        clearSuggestions,
    } = usePlacesAutocomplete({
        // some preference for search, totally optional
        requestOptions: {
            location: {
                lat: () => currLocation.lat,
                lng: () => currLocation.lng,
            },
            radius: 200 * 1000,
        },
    });

    async function handleSubmit(e) {
        e.preventDefault();
        let res = await addActivity(formData);
        if (res.success) {
            setIsAdding(false);
        } else {
            setFormErrors(res.errs);
        }
    }

    function handleCancel() {
        setIsAdding(false);
        setSelectLocation(null);
    }

    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData((data) => ({ ...data, [name]: value }));
    }

    return (
        <form
            onSubmit={handleSubmit}
            className="activity-add-form ml-3 mb-3 p-3 text-left"
        >
            <div className="form-group">
                <label htmlFor="activityName">*Activity Name: </label>
                <input
                    name="activityName"
                    id="activityName"
                    className="form-control"
                    value={formData.activityName}
                    onChange={handleChange}
                    autoComplete="off"
                    required
                    type="text"
                ></input>
            </div>

            <div className="form-group">
                <label htmlFor="description">Description: </label>
                <input
                    name="description"
                    id="description"
                    className="form-control"
                    value={formData.description}
                    onChange={handleChange}
                    autoComplete="off"
                    type="text"
                ></input>
            </div>

            <div className="form-group">
                <label htmlFor="location">Location: </label>
                <Combobox
                    onSelect={async (address) => {
                        setValue(address, false);
                        clearSuggestions();
                        try {
                            const results = await getGeocode({ address });
                            const { lat, lng } = await getLatLng(results[0]);
                            setFormData((data) => ({
                                ...data,
                                location: address,
                                latitude: lat,
                                longitude: lng,
                            }));
                            setSelectLocation({ lat: lat, lng: lng });
                        } catch (e) {
                            console.log(e);
                        }
                    }}
                >
                    <ComboboxInput
                        value={value}
                        className="form-control"
                        name="location"
                        id="location"
                        onChange={(e) => {
                            setValue(e.target.value);
                            handleChange(e);
                        }}
                        disabled={!ready}
                        type="text"
                        placeholder="Enter an address"
                        autoComplete="off"
                        required
                    />
                    <ComboboxPopover>
                        <ComboboxList>
                            {status === "OK" &&
                                data.map(({ id, description }) => (
                                    <ComboboxOption
                                        key={id}
                                        value={description}
                                    ></ComboboxOption>
                                ))}
                        </ComboboxList>
                    </ComboboxPopover>
                </Combobox>
            </div>

            <div className="form-group">
                <label htmlFor="startTime">Start Time: </label>
                <br />
                <TextField
                    id="startTime"
                    name="startTime"
                    type="datetime-local"
                    value={formData.startTime}
                    onChange={handleChange}
                />
            </div>

            <div className="form-group">
                <label htmlFor="endTime">End Time: </label>
                <br />
                <TextField
                    id="endTime"
                    name="endTime"
                    type="datetime-local"
                    value={formData.endTime}
                    onChange={handleChange}
                />
            </div>

            <div className="form-group">
                <label htmlFor="comment">Comment: </label>
                <input
                    name="comment"
                    id="comment"
                    className="form-control"
                    value={formData.comment}
                    onChange={handleChange}
                    type="text"
                ></input>
            </div>

            {formErrors.length ? (
                <ul>
                    {formErrors.map((e) => (
                        <li className="text-danger">{e}</li>
                    ))}
                </ul>
            ) : null}
            <div className="text-center">
                <button className="btn btn-success mb-3 mr-5" type="submit">
                    Save
                </button>
                <button
                    className="btn btn-danger btn-danger-trip mb-3"
                    onClick={handleCancel}
                >
                    Cancel
                </button>
            </div>
        </form>
    );
}

export default AddActivityForm;
