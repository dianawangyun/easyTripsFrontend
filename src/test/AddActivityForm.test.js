import React from "react";
import { render } from "@testing-library/react";
import AddActivityForm from "../activity/AddActivityForm";
import UserContext from "../auth/UserContext";

function renderAddActivityForm(currLocation) {
    return render(
        <UserContext.Provider value={currLocation}>
            <AddActivityForm />
        </UserContext.Provider>
    );
}

const currLocation = { lat: 30, lng: 100 };

it("renders without crashing", function () {
    renderAddActivityForm(currLocation);
});

it("matches snapshot", function () {
    const { asFragment } = renderAddActivityForm(currLocation);
    expect(asFragment()).toMatchSnapshot();
});

it("should show the form", function () {
    const { queryByText } = renderAddActivityForm(currLocation);
    expect(queryByText("*Activity Name:")).toBeInTheDocument();
});
