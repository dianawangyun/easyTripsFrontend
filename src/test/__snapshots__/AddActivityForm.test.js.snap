// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="activity-add-form ml-3 mb-3 p-3 text-left"
  >
    <div
      class="form-group"
    >
      <label
        for="activityName"
      >
        *Activity Name: 
      </label>
      <input
        autocomplete="off"
        class="form-control"
        id="activityName"
        name="activityName"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="location"
      >
        Description: 
      </label>
      <input
        class="form-control"
        id="description"
        name="description"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="location"
      >
        Location: 
      </label>
      <div
        data-reach-combobox=""
        data-state="idle"
      >
        <input
          aria-autocomplete="both"
          aria-controls="listbox--3"
          aria-expanded="false"
          aria-haspopup="listbox"
          autocomplete="false"
          class="form-control"
          data-reach-combobox-input=""
          data-state="idle"
          disabled=""
          id="location"
          name="location"
          placeholder="Enter an address"
          role="combobox"
          type="search"
          value=""
        />
      </div>
    </div>
    <div
      class="form-group"
    >
      <label
        for="startTime"
      >
        Start Time: 
      </label>
      <br />
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="startTime"
            name="startTime"
            type="datetime-local"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="form-group"
    >
      <label
        for="endTime"
      >
        End Time: 
      </label>
      <br />
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="endTime"
            name="endTime"
            type="datetime-local"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="form-group"
    >
      <label
        for="comment"
      >
        Comment: 
      </label>
      <input
        class="form-control"
        id="comment"
        name="comment"
        type="text"
        value=""
      />
    </div>
    <div
      class="text-center"
    >
      <button
        class="btn btn-success mb-3 mr-5"
        type="submit"
      >
        Save
      </button>
      <button
        class="btn btn-danger btn-danger-trip mb-3"
      >
        Cancel
      </button>
    </div>
  </form>
</DocumentFragment>
`;
